pkg/buildconfig/writer.go

```diff
@@ -0,0 +1,136 @@
+package buildconfig
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"go/format"
+	"io"
+	"os"
+	"sort"
+	"strings"
+	"text/template"
+
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
+	"gopkg.in/yaml.v3"
+)
+
+type GoConstantsWriter struct {
+	input  io.Reader
+	output io.Writer
+	buf    []byte
+	cfg    map[string]any
+	tmpl   *template.Template
+}
+
+// NewGoConstantsWriter creates a new transformer of read YAML into a block of Go constants.
+func NewGoConstantsWriter(t *template.Template, in io.Reader, out io.Writer) *GoConstantsWriter {
+	return &GoConstantsWriter{
+		tmpl:   t,
+		input:  in,
+		output: out,
+	}
+}
+
+// Run loads YAML data from the pre-configured input source, processes it, and outputs a template with formatted
+// Go constants in the pre-configured output source. This method can only be run once, since the input source gets fully read.
+func (f *GoConstantsWriter) Run() error {
+	if err := f.load(); err != nil {
+		return err
+	}
+	if err := f.process(); err != nil {
+		return err
+	}
+	if err := f.write(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (f *GoConstantsWriter) load() error {
+	if f.input == nil {
+		return errors.New("nil input")
+	}
+	b, err := io.ReadAll(f.input)
+	if err != nil {
+		return err
+	}
+	if len(b) == 0 {
+		return errors.New("nothing was read, the input source might be fully read")
+	}
```

Comment:
I am inclined to agree here. I just thought it's a special case of error that we want to bring to someone's attention - since each value of this type is to be used only once.